units/generate_tasks.go

```diff
@@ -47,11 +47,17 @@ func makeGenerateTaskJob() *generateTasksJob {
 	return j
 }
 
-func NewGenerateTasksJob(id string, ts string) amboy.Job {
+// NewGenerateTasksJob returns a job that dynamically updates the project
+// configuration based on the given task's generate.tasks configuration.
+func NewGenerateTasksJob(versionID, taskID string, ts string, useScopes bool) amboy.Job {
 	j := makeGenerateTaskJob()
-	j.TaskID = id
+	j.TaskID = taskID
 
-	j.SetID(fmt.Sprintf("%s-%s-%s", generateTasksJobName, id, ts))
+	j.SetID(fmt.Sprintf("%s-%s-%s", generateTasksJobName, taskID, ts))
+	if useScopes {
+		j.SetScopes([]string{versionID, taskID})
```

Comment:
I'm curious what this means; I've noticed that in some other places we use scopes we use "%s.%s" rather than passing in two strings -- is that functionally the same thing or is passing in the list different? How does this interact with SetEnqueueScopes?

Reply:
Each string counts as a separate lock. These locks are functionally the same as the `%s.%s` format in this case because generate task jobs are in their own dedicated queue. But for the sake of consistency, I changed it to use the `%s.%s` format including the job name.

The two locks achieve slightly different goals. `SetEnqueueScope` tells Amboy when to apply the lock in the job lifecycle. By default, a normal scope applies the lock when the job is dispatched, which prevents multiple jobs with the same scope from running at the same time. If a scope is marked as an enqueue scope, it applies the lock when the job is enqueued rather than when it's dispatched. Applying the scope earlier disallows multiple jobs with the same enqueue scope from being put in the queue. This resolves an issue where jobs that have to do the same work might end up being enqueued more than once. For example, in the current way generate tasks works (without enqueue scopes), if task ABC is running generate tasks, there's nothing stopping the cron populator from enqueueing multiple jobs that are effectively supposed to handle the same task except by adding a rounded timestamp to the `_id`. This deduplication u sing timestamps is not really necessary now that enqueue scopes are available. Using the task-level enqueue scope means that there can be at most 1 generate tasks job for ABC in the Amboy queue at any given time, either waiting to run or running.

Using the version-level scope satisfies the correctness goal of preventing multiple tasks in the same version from running concurrently. Using the task-level enqueue scope is a performance improvement to prevent duplicate generate task jobs from being enqueued for the same task.

Reply:
Very cool, thank you!